### PR TITLE
Add support to control output file behavior

### PR DESCRIPTION
**Pull request by javierluraschi**
Minor additions to control the behavior of output files:

- Read `profvis.output_path` option to define the default output location of profiles. This option can be set through `options(profvis.output_path = "<path>")`
- Read 'profvis.keep_output` to keep the temporary profile even after profvis closes. This enables further processing by the host of profvis. This option can be set through `options(profvis.keep_output = TRUE)`
- Return the `output_path` under the widget structure to allow the host to retrieve profile names generated by profvis. This data can be accessed with `profvis(pause(1))$x$message$prof_output`
- Fix to parsing profiles that contain no sources and no `<expr>`. Hit this corner case while running:
```
Rprof("~/tmp/test.rprof")
pause(1)
Rprof(NULL)
profvis(prof_input = "~/tmp/test.rprof")
```

### CODE REVIEW

**Comment by wch** [comment]
Looks pretty good! The only thing that I think we'd want to do differently relates to the `profvis.keep_output` option. I think that should also be an argument to the function. So the function signature could be something like this:

```R
 profvis <- function(expr = NULL, interval = 0.01, prof_output = NULL,
                     prof_input = NULL, keep_output = NULL,
                     width = NULL, height = NULL)
```

It would be nice if we could set the default in the function signature, as in `function(..., keep_output = !is.null(prof_output))`, but the logic is a bit too complicated if we want to respect the `profvis.output_path` option.

The new option will also need to be documented.

**Comment by javierluraschi** [comment]
This, conceptually, makes sense; however, took a look at the code and seems that by default we save the profiles in the `tmp` directory through `tempfile(fileext = ".prof")`. In my system, files end up being saved, for instance, as `/tmp/Rtmpocx46J/file7129119e25b4.prof`. This location happens to be cleared by OSX continuously and I'm loosing the profiles anyway, even if profvis does not clean them.

Seems to me that if we want to surface this functionality as a parameter, we would also want to surface the `output_path` parameter. However, if we go this route, it would make sense to assume that when the `output_path` is specified, users want to keep the files. If we do this, we could get rid of the `keep_output` parameter all together.

Thoughts?

**Comment by wch** [comment]
The temp dir is cleared every time you restart R, which I'm guessing you're doing fairly frequently.

The two arguments `prof_output` and `output_path` would interact in a confusing way. How about this:

* Keep the function args unchanged from the master branch.
* If `prof_output` is NULL, set it to `getOption("profvis.prof_output")`
* If `prof_output` is a directory, create a temp file in that directory; if it's a file, create that file.
* In RStudio, set `profvis.prof_output` to some directory.

**Comment by javierluraschi** [comment]
:+1:

**Comment by javierluraschi** [comment]
@wch: looks much better, consistent and simple.

File: R/profvis.R [hunk 1]
```diff
@@ -11,9 +11,9 @@
 #' @param expr Code to profile. Not compatible with \code{prof_input}.
 #' @param interval Interval for profiling samples, in seconds. Values less than
 #'   0.005 (5 ms) will probably not result in accurate timings
-#' @param prof_output Name of an Rprof output file in which to save profiling
-#'   data. If \code{NULL} (the default), a temporary file will be used and
-#'   automatically removed when the function exits.
+#' @param prof_output Name of an Rprof output file or directory in which to save
```
Comment (wch) [inline R/profvis.R]: Could you mention that if it's a directory, a random filename will be used?

**Comment by javierluraschi** [comment]
For sure! Updated.

**Comment by wch** [comment]
Thanks!
